### ISSUE TITLE
Reassign block requests for immediate download, after peer returns short response.

### ISSUE THREAD

**Issue by carver** [ethereum/py-evm#879]
More discussion at #868 

Roughly:

```
Request blocks 1-5 from Peer A
Request blocks 6-10 from Peer B
Peer A responds with blocks 1-2
<-- at this point, we can request blocks 3-5 from Peer A, while waiting for Peer B to respond
```

**Comment by pipermerriam** [comment]
@carver so re-request from same peer who responded with less information than we requested?

**Comment by carver** [comment]
Yup, that was the idea. Returning a short result isn't the sign of misbehaving or a bad peer. They probably just have a byte size limit, and returned quickly once they hit that limit.

I think the best way to solve it, though, is a refactor of how the split and send works: to separately track the desired headers and the peers with outstanding requests, rather than batch and wait to re-request from all peers at once.

Anyway, I'm hesitant to spend much time optimizing without performance data, so I'll probably wrap #1009 as quickly as reasonable and leave this issue for sometime later.
